File: src-react/src/lib/query/sync.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { syncApi } from "@/lib/api/sync";
import type { AddTaskInput, CreateTaskGroupInput } from "@/lib/api/sync";
import type { SessionBackup, Task, TaskGroup } from "@/types";

export const syncKeys = {
  webdavSettings: ["sync", "webdavSettings"] as const,
  taskGroups: ["sync", "taskGroups"] as const,
  sessionBackups: ["sync", "sessionBackups"] as const,
};

export function useWebdavSettingsQuery() {
  return useQuery({
    queryKey: syncKeys.webdavSettings,
    queryFn: syncApi.getWebdavSettings,
  });
}

export function useSaveWebdavSettingsMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: syncApi.saveWebdavSettings,
    onSuccess: (settings) => {
      queryClient.setQueryData(syncKeys.webdavSettings, settings);
    },
  });
}

export function useTestWebdavConnectionMutation() {
  return useMutation({
    mutationFn: syncApi.testWebdavConnection,
  });
}

export function useTaskGroupsQuery() {
  return useQuery({
    queryKey: syncKeys.taskGroups,
    queryFn: syncApi.listTaskGroups,
  });
}

export function useSessionBackupsQuery() {
  return useQuery({
    queryKey: syncKeys.sessionBackups,
    queryFn: syncApi.listSessionBackups,
    refetchOnMount: "always",
  });
}

export function useCreateTaskGroupMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: CreateTaskGroupInput) => syncApi.createTaskGroup(input),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: syncKeys.taskGroups }),
  });
}

export function useDeleteTaskMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: string) => syncApi.deleteTask(id),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: syncKeys.taskGroups }),
  });
}

export function useDeleteTaskGroupMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: string) => syncApi.deleteTaskGroup(id),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: syncKeys.taskGroups }),
  });
}

export function useAddTaskMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ groupId, task }: { groupId: string; task: AddTaskInput }) =>
      syncApi.addTask(groupId, task),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: syncKeys.taskGroups }),
  });
}

export function useUpdateTaskScheduleMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ id, schedule }: { id: string; schedule: string }) =>
      syncApi.updateTaskSchedule(id, schedule),
    onSuccess: (updated) => {
      queryClient.setQueryData<TaskGroup[]>(syncKeys.taskGroups, (groups) =>
        replaceTask(groups, updated),
      );
      queryClient.setQueryData<SessionBackup[]>(syncKeys.sessionBackups, (backups) =>
        replaceSessionBackupTask(backups, updated),
      );
    },
  });
}

export function useUpdateGroupScheduleMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ groupId, schedule }: { groupId: string; schedule: string }) =>
      syncApi.updateGroupSchedule(groupId, schedule),
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: syncKeys.taskGroups });
      void queryClient.invalidateQueries({ queryKey: syncKeys.sessionBackups });
    },
  });
}

export function useRunTaskMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: string) => syncApi.runTask(id),
    onSuccess: (updated) => {
      queryClient.setQueryData<TaskGroup[]>(syncKeys.taskGroups, (groups) =>
        replaceTask(groups, updated),
      );
      queryClient.setQueryData<SessionBackup[]>(syncKeys.sessionBackups, (backups) =>
        replaceSessionBackupTask(backups, updated),
      );
    },
  });
}

export function useReorderTaskGroupsMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (groupIds: string[]) => syncApi.reorderTaskGroups(groupIds),
    onSuccess: (groups) => {
      queryClient.setQueryData<TaskGroup[]>(syncKeys.taskGroups, groups);
    },
  });
}

export function useReorderTasksMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ groupId, taskIds }: { groupId: string; taskIds: string[] }) =>
      syncApi.reorderTasks(groupId, taskIds),
    onSuccess: (group) => {
      queryClient.setQueryData<TaskGroup[]>(syncKeys.taskGroups, (groups) =>
        groups?.map((g) => (g.id === group.id ? group : g)),
      );
    },
  });
}

function replaceTask(groups: TaskGroup[] | undefined, updated: Task): TaskGroup[] | undefined {
  return groups?.map((group) => ({
    ...group,
    tasks: group.tasks.map((task) => (task.id === updated.id ? updated : task)),
  }));
}

function replaceSessionBackupTask(
  backups: SessionBackup[] | undefined,
  updated: Task,
): SessionBackup[] | undefined {
  return backups?.map((backup) =>
    backup.task.id === updated.id ? { ...backup, task: updated } : backup,
  );
}

File: src-tauri/src/commands/sync.rs
use tauri::State;

use nexus_core::{
    error::AppResult,
    services::sync::{
        CreateTaskGroupInput, CreateTaskInput, SessionBackup, Task, TaskGroup, WebdavSettings,
        WebdavSettingsInput,
    },
};

use crate::store::AppState;

#[tauri::command]
pub fn get_webdav_settings(state: State<'_, AppState>) -> AppResult<WebdavSettings> {
    state.sync.get_webdav_settings()
}

#[tauri::command]
pub fn save_webdav_settings(
    state: State<'_, AppState>,
    input: WebdavSettingsInput,
) -> AppResult<WebdavSettings> {
    state.sync.save_webdav_settings(input)
}

#[tauri::command]
pub async fn test_webdav_connection(
    state: State<'_, AppState>,
    input: WebdavSettingsInput,
) -> AppResult<()> {
    state.sync.test_webdav_connection(input).await
}

#[tauri::command]
pub fn list_task_groups(state: State<'_, AppState>) -> AppResult<Vec<TaskGroup>> {
    state.sync.list_task_groups()
}

#[tauri::command]
pub fn list_session_backups(state: State<'_, AppState>) -> AppResult<Vec<SessionBackup>> {
    state.sync.list_session_backups()
}

#[tauri::command]
pub fn create_task_group(
    state: State<'_, AppState>,
    input: CreateTaskGroupInput,
) -> AppResult<TaskGroup> {
    state.sync.create_task_group(input)
}

#[tauri::command]
pub fn delete_task(state: State<'_, AppState>, id: String) -> AppResult<()> {
    state.sync.delete_task(id)
}

#[tauri::command]
pub fn delete_task_group(state: State<'_, AppState>, id: String) -> AppResult<()> {
    state.sync.delete_task_group(id)
}

#[tauri::command]
pub fn add_task(
    state: State<'_, AppState>,
    group_id: String,
    task: CreateTaskInput,
) -> AppResult<TaskGroup> {
    state.sync.add_task(group_id, task)
}

#[tauri::command]
pub fn update_task_schedule(
    state: State<'_, AppState>,
    id: String,
    schedule: String,
) -> AppResult<Task> {
    state.sync.update_task_schedule(id, schedule)
}

#[tauri::command]
pub fn update_group_schedule(
    state: State<'_, AppState>,
    group_id: String,
    schedule: String,
) -> AppResult<()> {
    state.sync.update_group_schedule(group_id, schedule)
}

#[tauri::command]
pub fn reorder_task_groups(
    state: State<'_, AppState>,
    group_ids: Vec<String>,
) -> AppResult<Vec<TaskGroup>> {
    state.sync.reorder_task_groups(group_ids)
}

#[tauri::command]
pub fn reorder_tasks(
    state: State<'_, AppState>,
    group_id: String,
    task_ids: Vec<String>,
) -> AppResult<TaskGroup> {
    state.sync.reorder_tasks(group_id, task_ids)
}

#[tauri::command]
pub async fn run_task(state: State<'_, AppState>, id: String) -> AppResult<Task> {
    state.sync.run_task(id).await
}
